File: hw2/java/Restaurant.java
public class Restaurant {

    private static final int NUM_CHEFS = 3;
    private static final int NUM_WAITERS = 2;
    private static final int NUM_PHILOSOPHERS = 5;

    private final Table table;
    private final Kitchen kitchen;
    private Chef[] chefs;
    private Waiter[] waiters;
    private Philosopher[] philosophers;
    private final String[] PHILOSOPHER_NAMES = {
        "Plato",
        "Socrates",
        "Aristotle",
        "Dostoevsky",
        "Marcus Aurelius",
    };

    private static final String[] WAITER_NAMES = { "Miria ", "Isaac" };

    private static final String[] CHEF_NAMES = { "Eren", "Mikasa", "Armin" };

    public Restaurant(Table table, Kitchen kitchen) {
        this.table = table;
        this.table.setRestaurant(this);
        this.kitchen = kitchen;
        this.kitchen.setRestaurant(this);

        this.chefs = new Chef[NUM_CHEFS];
        for (int i = 0; i < this.chefs.length; i++) {
            this.chefs[i] = new Chef(this.kitchen, CHEF_NAMES[i]);
        }

        this.waiters = new Waiter[NUM_WAITERS];
        for (int i = 0; i < this.waiters.length; i++) {
            this.waiters[i] = new Waiter(
                this.table,
                this.kitchen,
                WAITER_NAMES[i]
            );
        }

        this.philosophers = new Philosopher[NUM_PHILOSOPHERS];
        for (int i = 0; i < this.philosophers.length; i++) {
            this.philosophers[i] = new Philosopher(
                this.table,
                PHILOSOPHER_NAMES[i]
            );
        }
    }

    public void start() {
        for (Philosopher philosopher : this.philosophers) {
            philosopher.start();
        }

        for (Chef chef : this.chefs) {
            chef.start();
        }

        for (Waiter waiter : this.waiters) {
            waiter.start();
        }
    }

    public void shutdown() {
        for (Philosopher philosopher : this.philosophers) {
            try {
                philosopher.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        for (Waiter waiter : this.waiters) {
            waiter.shutdown();
        }
        for (Waiter waiter : this.waiters) {
            try {
                waiter.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        for (Chef chef : this.chefs) {
            chef.shutdown();
        }
        for (Chef chef : this.chefs) {
            try {
                chef.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public synchronized Chef getRandomChef() {
        int randomIndex = (int) (Math.random() * this.chefs.length);
        Chef chef = this.chefs[randomIndex];
        if (chef.isBusy()) {
            return null;
        }
        return chef;
    }

    public synchronized Waiter getRandomWaiter() {
        int randomIndex = (int) (Math.random() * this.waiters.length);
        Waiter waiter = this.waiters[randomIndex];
        if (waiter.isBusy()) {
            return null;
        }
        return waiter;
    }

    public Philosopher getPhilosopher(int seatNo) {
        return this.philosophers[seatNo];
    }

    public String getPhilosopherName(int seatNo) {
        return this.philosophers[seatNo].name();
    }
}

File: hw2/go/restaurant.go
package main

import (
	"math/rand"
	"sync"
)

const (
	NUM_PHILOSOPHER      uint = 5
	NUM_WAITERS          uint = 2
	NUM_SEATS            uint = NUM_PHILOSOPHER
	NUM_CHOPSTICKS       uint = NUM_SEATS
	KITCHEN_COUNTER_SIZE uint = NUM_PHILOSOPHER
	NUM_CHEFS            uint = 3
)

var PHIL_NAMES = []string{"Phil 1", "Phil 2", "Phil 3", "Phil 4", "Phil 5"}
var WAITER_NAMES = []string{"Waiter 1", "Waiter 2"}
var CHEF_NAMES = []string{"Chef 1", "Chef 2", "Chef 3"}

type Restaurant struct {
	philosophers   []*Philosopher
	waiters        []*Waiter
	chefs          []*Chef
	kitchenCounter chan *Order
	table          *Table
}

func newRestaurant() *Restaurant {
	log("Building Restaurant")
	r := &Restaurant{
		philosophers:   make([]*Philosopher, NUM_PHILOSOPHER),
		waiters:        make([]*Waiter, NUM_WAITERS),
		chefs:          make([]*Chef, NUM_CHEFS),
		kitchenCounter: make(chan *Order, KITCHEN_COUNTER_SIZE),
		table:          newTable(),
	}

	r.initWaiters()
	r.initChefs()
	r.initPhilosophers()

	return r
}

func (r *Restaurant) initPhilosophers() {
	for i := range NUM_PHILOSOPHER {
		r.philosophers[i] = newPhilosopher(PHIL_NAMES[i], r)
	}
	log("philosophers ready")
}

func (r *Restaurant) initWaiters() {
	for i := range r.waiters {
		r.waiters[i] = newWaiter(WAITER_NAMES[i], r)
	}
	log("Hiring waiters")
}

func (r *Restaurant) initChefs() {
	for i := range NUM_CHEFS {
		r.chefs[i] = newChef(CHEF_NAMES[i], r)
	}
	log("Hiring chefs")
}

func (r *Restaurant) giveSeat() uint {
	return r.table.giveSeat()
}

func (r *Restaurant) takeSeatBack(seat uint) {
	r.table.takeSeatBack(seat)
}

func (r *Restaurant) waitGetChopsticks(seat uint) {
	r.table.waitGetChopsticks(seat)
}

func (r *Restaurant) releaseChopsticks(seat uint) {
	r.table.releaseChopsticks(seat)
}

func (r *Restaurant) getRandomPhilosopher() *Philosopher {
	phil := r.philosophers[rand.Intn((int)(NUM_PHILOSOPHER))]
	assert(phil != nil)
	return phil
}

func (r *Restaurant) getRandomChef() *Chef {
	chef := r.chefs[rand.Intn(len(r.chefs))]
	assert(chef != nil)
	return chef
}

func (r *Restaurant) putOrderOnKitchenCounter(order *Order) {
	r.kitchenCounter <- order
}

func (r *Restaurant) terminateChefs() {
	for i := range r.chefs {
		r.chefs[i].terminate()
	}
}

func (r *Restaurant) terminateWaiters() {
	for i := range r.waiters {
		r.waiters[i].terminate()
	}
}

func (r *Restaurant) run() {
	pwg := sync.WaitGroup{}
	wg := sync.WaitGroup{}
	pwg.Add((int)(NUM_PHILOSOPHER))
	wg.Add((int)(NUM_WAITERS + NUM_CHEFS))
	log("Restaurant is now open for business")
	for i := range r.philosophers {
		go func(p *Philosopher) {
			assert(p != nil)
			defer pwg.Done()
			p.run()
		}(r.philosophers[i])
	}

	for i := range r.waiters {
		go func(w *Waiter) {
			assert(w != nil)
			defer wg.Done()
			w.run()
		}(r.waiters[i])
	}

	for i := range r.chefs {
		go func(c *Chef) {
			assert(c != nil)
			defer wg.Done()
			c.run()
		}(r.chefs[i])
	}
	pwg.Wait()
	r.terminateChefs()
	r.terminateWaiters()
	wg.Wait()
}
